File: onyx-database/src/main/java/com/onyx/diskmap/base/hashmap/AbstractIterableHashMap.java
package com.onyx.diskmap.base.hashmap;

import com.onyx.diskmap.node.Header;
import com.onyx.diskmap.store.Store;

import java.util.AbstractCollection;
import java.util.Collection;
import java.util.Iterator;

/**
 * This class maintains the iterator capablities for the hash table.
 *
 * @param <K> Key
 * @param <V> Value
 */
abstract class AbstractIterableHashMap<K, V> extends AbstractCachedHashMap<K, V> {

    /**
     * Constructor
     *
     * @param fileStore File storage
     * @param header Head of the data structore
     * @param headless Whether it is used in a stateless manner.
     * @param loadFactor Indicates how large to allocate the initial data structure
     *
     * @since 1.2.0
     */
    AbstractIterableHashMap(Store fileStore, Header header, boolean headless, int loadFactor) {
        super(fileStore, header, headless, loadFactor);
    }

    /**
     * Collection of skip list references
     * @return Collection of skip list references
     */
    Collection skipListMaps() {
        return new SkipListMapSet();
    }

    /**
     * Iterates through the skip list references.
     */
    private class SkipListMapIterator implements Iterator {
        int index = 0;

        @Override
        public boolean hasNext() {
            return index < getMapCount();
        }

        @Override
        @SuppressWarnings("unchecked")
        public Object next() {
            try {
                int mapId = getMapIdentifier(index);
                return getReference(mapId);
            }
            finally {
                index ++;
            }
        }
    }

    private class SkipListMapSet extends AbstractCollection {
        @Override
        public Iterator iterator() {
            return new SkipListMapIterator();
        }

        @Override
        public int size() {
            return getMapCount();
        }
    }
}

File: onyx-database/src/main/kotlin/com/onyx/diskmap/impl/base/hashmap/AbstractIterableHashMap.kt
package com.onyx.diskmap.impl.base.hashmap

import com.onyx.diskmap.data.Header
import com.onyx.diskmap.store.Store

import java.util.AbstractCollection

/**
 * This class maintains the iterator capabilities for the hash table.
 *
 * @param <K> Key
 * @param <V> Value
 */
@Suppress("UNCHECKED_CAST")
abstract class AbstractIterableHashMap<K, V>(fileStore: Store, header: Header, headless: Boolean, loadFactor: Int) : AbstractCachedHashMap<K, V>(fileStore, header, headless, loadFactor) {

    /**
     * Collection of skip list references
     * @return Collection of skip list references
     */
    internal val skipListMaps: Collection<Long>
        get() = SkipListMapSet()

    /**
     * Iterates through the skip list references.
     */
    private inner class SkipListMapIterator : MutableIterator<Long> {
        override fun remove() {}

        internal var index = 0

        override fun hasNext(): Boolean = index < mapCount.get()

        override fun next(): Long {
            try {
                val mapId = getSkipListIdentifier(index)
                return getSkipListReference(mapId)
            } finally {
                index++
            }
        }
    }

    private inner class SkipListMapSet<T> : AbstractCollection<T>() {
        override fun iterator(): MutableIterator<T> = SkipListMapIterator() as MutableIterator<T>
        override val size: Int
            get() = mapCount.get()
    }
}
